custom-metrics-stackdriver-adapter/pkg/adapter/provider/translator.go

```diff
@@ -0,0 +1,163 @@
+	if err != nil {
+		return nil, err
+	}
+	if len(values) != 1 {
+		return nil, fmt.Errorf("Expected exactly one value for pod %s in namespace %s, but received %v values", name, namespace, len(values))
+	}
+	for _, value := range values {
+		metricValue, err := t.metricFor(value, groupResource, namespace, name, metricName)
+		if err != nil {
+			return nil, err
+		}
+		return metricValue, nil
+	}
+	// This code is unreacheable
+	return nil, fmt.Errorf("Illegal state")
+}
+
+func (t *Translator) createFilter(objectIDs []string, project string, zone string, clusterName string, metricName string) (string, error) {
+	// project_id, cluster_name and pod_id together identify a pod unambiguously
+	projectFilter := fmt.Sprintf("resource.label.project_id = \"%s\"", project)
+	zoneFilter := fmt.Sprintf("resource.label.zone = \"%s\"", zone)
+	clusterFilter := fmt.Sprintf("resource.label.cluster_name = \"%s\"", strings.TrimSpace(clusterName))
+	// container_name is set to empty string for pod metrics
+	containerFilter := "resource.label.container_name = \"\""
+	metricFilter := fmt.Sprintf("metric.type = \"%s/%s\"", t.config.MetricsPrefix, metricName)
+
+	var nameFilter string
+	if len(objectIDs) == 0 {
+		return "", fmt.Errorf("No pods matched for metric %s", metricName)
+	} else if len(objectIDs) == 1 {
+		nameFilter = fmt.Sprintf("resource.label.pod_id = %s", objectIDs[0])
+	} else {
+		nameFilter = fmt.Sprintf("resource.label.pod_id = one_of(%s)", strings.Join(objectIDs, ","))
+	}
+	return fmt.Sprintf("(%s) AND (%s) AND (%s) AND (%s) AND (%s) AND (%s)", metricFilter, projectFilter, clusterFilter, zoneFilter, nameFilter, containerFilter), nil
+}
+
+func (t *Translator) getMetricValuesFromResponse(groupResource schema.GroupResource, namespace string, response *stackdriver.ListTimeSeriesResponse) (map[string]int64, error) {
+	group, err := api.Registry.Group(groupResource.Group)
+	if err != nil {
+		return nil, err
+	}
+	_, err = api.Registry.RESTMapper().KindFor(groupResource.WithVersion(group.GroupVersion.Version))
+	if err != nil {
+		return nil, err
+	}
+	if len(response.TimeSeries) < 1 {
+		return nil, fmt.Errorf("Expected at least one time series from Stackdriver, but received %v", len(response.TimeSeries))
+	}
+	metricValues := make(map[string]int64)
+	// Find time series with specified labels matching
+	// Stackdriver API doesn't allow complex label filtering (i.e. "label1 = x AND (label2 = y OR label2 = z)"),
+	// therefore only part of the filters is passed and remaining filtering is done here.
+	for _, series := range response.TimeSeries {
+		if len(series.Points) != 1 {
+			return nil, fmt.Errorf("Expected exactly one Point in TimeSeries from Stackdriver, but received %v", len(series.Points))
+		}
+		value := *series.Points[0].Value
+		name := series.Resource.Labels["pod_id"]
+		switch {
+		case value.Int64Value != nil:
+			metricValues[name] = *value.Int64Value
+		case value.DoubleValue != nil:
+			metricValues[name] = int64(*value.DoubleValue)
```

Comment:
instead of just discarding the decimal component, I'd construct this as a milli-quantity.